src/Modules/OrchardCore.Commerce/Services/ShoppingCartHelpers.cs

```diff
@@ -120,13 +122,35 @@ public class ShoppingCartHelpers : IShoppingCartHelpers
             totals = lines.CalculateTotals().ToList();
         }
 
-        // The values are rounded to avoid storing more precision than what the currency supports.
+        // Checkout should not be possible if any of the items are unpurchasable.
```

Comment:
The model we are dealing with here is `ShoppingCartViewModel`, while inside `ViewModelCreatedAsync()` it's the `CheckoutViewModel` that's handled (after making the above requested change to pass in `ICheckoutViewModel`).

I suppose both could be passed in as nullable and then the method can handle them based on which one is not null. Though there might be a more elegant way to do this?

```
    public Task ViewModelCreatedAsync(
        IList<ShoppingCartLineViewModel> lines,
        ICheckoutViewModel checkoutViewModel = null,
        ShoppingCartViewModel shoppingCartViewModel = null)
```

Reply:
OCC is so complex, we really can't afford code duplication of hard assumptions about which features are enabled. So it's important to treat features like Inventory as plugins instead of wiring them into the main program. In this case you could do the following:
1. Move the business logic from InventoryCheckoutEvents into a service (maybe `ProductInventoryService`?) using your original method signature, i.e. `Task<bool> VerifyLinesAsync(IList<ShoppingCartLineViewModel> lines)`.
2. Add a `ViewModelCreatedAsync` method to `IShoppingCartEvents` as well.
3. Now both `InventoryCheckoutEvents.ViewModelCreatedAsync()` and `InventoryShoppingCartEvents.ViewModelCreatedAsync()` can call the service with the shared logic and update their respective view-models without harming future extension possibilities.